x-pack/plugins/security/public/session_expired.test.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { BasePath } from 'src/core/public/http/base_path_service';
```

Comment:
@azasypkin @legrego how do you all feel about us writing tests using the NP core services, instead of mocking them out? When it's painful, or impossible, to use a concrete instance, I definitely think it makes sense to use a mock. However, in situations like these it really isn't that painful, and seems to add a level of additional confidence that the code actually works.

Reply:
I'm in favor of that approach. I prefer concrete instances when it's not too painful to create or maintain, for the same reasons as you.

Reply:
Both approaches sound good to me too at this stage, but only because of `additional confidence that the code actually works` point. I'm afraid using real services may not scale very well and we'll end up with a mock in one test and real service in another.

The issue here is that not all core services are exposed to plugins as is, moreover they may have different shape for different life-cycle events: what you can do with HTTP service in `setup` may be drastically different from what you can do with it in `start` (e.g. [this](https://github.com/elastic/kibana/blob/c9220c586306ae7082456ec628c4a919f66a0cde/src/core/server/plugins/plugin_context.ts#L110-L118) or [this](https://github.com/elastic/kibana/blob/b94c16da75bf7fc414e7f982da33da9ee83aadee/src/core/public/plugins/plugin_context.ts#L79)). That means that some of the core services won't be "public" (btw, if I'm not mistaken we're only supposed to import from `src/core/public` or `src/core/public/mocks`, I'm surprised that linter doesn't complain about `src/core/public/http/base_path_service`).

Having said that, core **should** provide mocks for all of its services and keep them in sync. If we don't have a mock for the core service that is used in a plugin or it's hard to use it in our tests (e.g. it doesn't have reasonable defaults for mock return values) - platform team should treat it as a bug :slightly_smiling_face:

Reply:
> I'm afraid using real services may not scale very well and we'll end up with a mock in one test and real service in another.

I share this fear as well. It seems like a lot of the existing NP tests use the mocks which the NP provides. 

> The issue here is that not all core services are exposed to plugins as is, moreover they may have different shape for different life-cycle events: what you can do with HTTP service in setup may be drastically different from what you can do with it in start (e.g. this or this). That means that some of the core services won't be "public" (btw, if I'm not mistaken we're only supposed to import from src/core/public or src/core/public/mocks, I'm surprised that linter doesn't complain about src/core/public/http/base_path_service).

It definitely felt like I was reaching "too far" into `/src/core/public` to get access to this. Do you happen to have a recommendation for where I should be getting the core mock from?

Reply:
I believe we're supposed to get all the mocks from either `src/core/server/mocks` or `src/core/public/mocks`. So for `base_path` it'd be something like this I guess:
```ts
import { coreMock } from 'src/core/public/mocks';
const basePath = coreMock.createSetup().http.basePath;
```

Reply:
Thanks @azasypkin! I'll change this to use the mocks and reevaluate the other tests as well

Reply:
I just re-wrote the tests for this using only the mocks, and it's significantly "uglier" and harder to understand... @joshdover how do you feel about us using a concrete instance of `BasePath` for tests like these?

Reply:
The basePath APIs are so small and don't interact with anything meaningful. It seems to me that we should replace this with the real implementation from `coreMock` and make `coreMock` configurable so you can pass in a basePath when you create the mock.

Normally, I'd say it's bad to mix concrete implementations and mocks like this, but given how little code we're talking about and how common I think this scenario is, I think this mixing would be fine.

In summary: delete the basePath mock completely, replace it with the real implementation, allow coreMock to take a basePath argument.